### ISSUE TITLE
Any way to retrieve all validation errors?

### ISSUE THREAD

**Issue by AlexKovalevych** [hrzndhrn/xema#117]
Currently i expect to have all validation errors at once, but i didn't get additional properties validation error unless i make all other fields valid. E.g.:

```
schema = %Xema{
  refs: %{
    "#/definitions/uuid" => %Xema.Schema{
      pattern: ~r/^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/,
      type: :string
    }
  },
  schema: %Xema.Schema{
    additional_properties: false,
    definitions: %{
      "uuid" => %Xema.Schema{
        pattern: ~r/^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/,
        type: :string
      }
    },
    properties: %{
      "foo" => %Xema.Schema{
        ref: %Xema.Ref{pointer: "#/definitions/uuid"}
      }
    },
    required: #MapSet<["foo"]>,
    schema: "http://json-schema.org/draft-04/schema#",
    type: :map
  }
}

Xema.validate(schema, %{"invalid" => "value", "foo" => "invalid_uuid"})

{:error,
 %Xema.ValidationError{
   message: nil,
   reason: %{
     properties: %{
       "foo" => %{
         pattern: ~r/^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/,
         value: "invalid_uuid"
       }
     }
   }
 }}

```

Once i fix `foo` value to be a valid uuid, i get additional property error:

```
Xema.validate(schema, %{
      "invalid" => "foo",
      "foo" => "0ffc467-061e-49f0-b544-cc4634213876"
})

{:error,
 %Xema.ValidationError{
   message: nil,
   reason: %{properties: %{"invalid" => %{additional_properties: false}}}
 }}
```

So, is there a way to get all of them within a single validation function call?

**Comment by NickNeck** [comment]
Unfortunately, this does not work at the moment. In some cases, errors will be collected but not in this case.

**Comment by jsmestad** [comment]
@NickNeck think this is something that can be added or is it not possible due to a design choice xema has?

**Comment by NickNeck** [comment]
@jsmestad that is on my todo list. Currently, I am working on another project, on my sailing license, and planning my vacation. But I will address this issue immediately afterward. Or do you need this feature really urgent?

**Comment by jsmestad** [comment]
The interest was because I was unsure if this was something that was fixable, and maybe something I could contribute. I ended up with Ecto Changeset's for the particular problem I was working on, but still interested in helping out if things need fixing :)

**Comment by NickNeck** [comment]
Hello @jsmestad , I have taken a look into the code and I have an idea to solve it.
The parts to change are the `validate_by` functions like this one: [lib/xema/validator.ex:132](https://github.com/hrzndhrn/xema/blob/master/lib/xema/validator.ex#L132)
I think the `with` should be refactored to not break up the validation and collect the errors. It could be looked like this:
```elixir
  defp validate_by(:map, schema, value, opts) do
    value
    |> Collector.new()
    |> Collector.collect(&size(schema, &1))
    |> Collector.collect(&keys(schema, &1))
    |> Collector.collect(&required(schema, &1))
    ...
    |> Collector.errors(fn errors ->
      merge(errors)
    end)
  end
```
In this case, the collector would handle the return values `:ok`, `{:error, reason}`, and `value` (updates the value for the next collect call). This will cause also a refactoring for `patterns/3`, `properties/3`, and `additionals/3`. 

Those are just a few weird thoughts, late at night. But maybe you have another idea.

**Comment by NickNeck** [comment]
This issue is partly solved in the next release (0.12.0).
Suppose you have the following schema:
```elixir
iex(1)> schema = Xema.new({
...(1)>   :map,
...(1)>   properties: %{foo: :integer},
...(1)>   pattern_properties: %{~r/str_.*/ => :string},
...(1)>   max_properties: 3,
...(1)>   additional_properties: false
...(1)> })
```
Then you get errors for `properties`, `pattern_properties`, and `addition_properties`: 
```elixir
iex(2)> Xema.validate(schema, %{foo: "foo", str_a: 5, add: :more})
{:error,
 %Xema.ValidationError{
   message: nil,
   reason: %{
     properties: %{
       add: %{additional_properties: false},
       foo: %{type: :integer, value: "foo"},
       str_a: %{type: :string, value: 5}
     }
   }
 }}
```
Before `0.12.0` you just get one of these errors. 

But if an "earlier" validation fails some errors are not be shown.
```elixir
iex(3)> Xema.validate(schema, %{foo: "foo", str_a: 5, add: :more, four: 4})
{:error,
 %Xema.ValidationError{
   message: nil,
   reason: %{
     max_properties: 3,
     value: %{add: :more, foo: "foo", four: 4, str_a: 5}
   }
 }}
```
